Add realistic tank-drive model

virtuald opened the pull request:
Tested it with robot code and some logging data from one of the robots I work with, and it's definitely better than what we have now. The output data is even vaguely believable.

The plan is to convert all existing physics examples to use this model, and perhaps get rid of the old drivetrain stuff in the future.

Based on paper at https://www.chiefdelphi.com/media/papers/3402 -- definitely wouldn't have been possible without their effort!

virtuald commented:
One thing that's a bit weird here is that the model uses metric units internally... but since everything else in the sim uses imperial units, I feel like it's 'better' to continue using them here and just do the conversions, instead of having this weird mixed-unit thing.

I suppose we could change the sim to be all metric units next season?

calcmogul commented on lib/pyfrc/physics/tankmodel.py:
```diff
@@ -0,0 +1,233 @@
+        self.l_position = 0
+        self.r_position = 0
+        
+        # this uses the formula for the moment of inertial for a circle...
```
"moment of inertial" -> "moment of inertia"

calcmogul commented on lib/pyfrc/physics/tankmodel.py:
```diff
@@ -0,0 +1,233 @@
+        rotation = (l - r) / float(self._x_wheelbase)
+        
+        velocity *= M_TO_FEET
+        rotation *= self.inertia
```
Multiplying by moment of inertia here makes this function return angular momentum, not angular velocity. Angular momentum `L = J * omega`.

calcmogul replied:
By doing conservation of angular momentum (sum of angular momenta = J*omega), I get the following. Let `b` be the robot's wheelbase over 2, `m` be the mass, `l` and `r` be the velocities of each side of the robot, `J` be the moment of inertia, and `omega` be the angular velocity of the robot.
```
-b * m * l + b * m * r = J * omega
b * m * (r - l) = J * omega
omega = b * m * (r - l) / J
```

virtuald replied:
That equation makes the robot go crazy :(

virtuald replied:
Converting to this equation is more realistic, and switching to a rectangular moment of inertia fixed the craziness issues.

virtuald commented:
FYI, there's an issue with the model and small masses... so let's not merge until that's resolved.

calcmogul commented on lib/pyfrc/physics/tankmodel.py:
```diff
@@ -0,0 +1,263 @@
+               x_wheelbase: float=_kitbot_wheelbase,
+               robot_width: float=_kitbot_width,
+               robot_length: float=_kitbot_length,
+               wheel_diameter: float=0.5,
```
Have you guys considered having Travis run a formatter to check for these things? WPILib uses yapf, which supports pep8, but it makes the args with type hints harder to read for this file because it puts several args on the same line. I'm pretty sure other formatters exist though.

virtuald replied:
It keeps getting suggested, but nobody seems to want to do it, and I don't care enough to do it. :)

calcmogul commented:
We figured out what the problem was. For small masses, the acceleration was high enough that the model would go unstable for large enough timesteps. This is an issue inherent to discretizing a continuous model. When you discretize a system, the location of the poles is distorted due to the phase loss introduced by the sample delay.  When the sample delay is large enough, the distortion can make a model that is otherwise always stable become unstable.

Forward Euler integration with `v = a * dt` is one type of discretization, although it has a lot of distortion so it can't be used for low sample rates (we saw instability with periods above 0.005). Zero-order hold is better in this regard. You can use Python Control to do the conversion for you, but you need the model in state-space form. Here's an example using the motor model in this PR: https://gist.github.com/calcmogul/9bc48948d8a250eb8963a6d3c93e98b9

I can write up a derivation for that state-space model if you'd like.

virtuald commented:
Hm. That example requires control and numpy... so we'd have to add those to pyfrc's requirements. numpy is no longer the worst thing in the world, since they distribute wheels for everything nowadays. It's standalone, which I like.

However, the control library requires numpy, scipy, and matplotlib along with a bunch of other stuff too? That's adding a ton of dependencies + indirect dependencies. Though, I suppose pytest brings in a ton of dependencies these days too.

Using matrix math etc is the right way to do this, but I'm wary of making the pyfrc inaccessible, and I wouldn't want to not include those in the requirements, otherwise the user will have to install yet another thing to figure out how to make this work.

Will have to contemplate. Won't have time tonight.

auscompgeek commented on lib/pyfrc/physics/units.py:
```diff
@@ -0,0 +1,25 @@
+M_TO_FT = 3.2808398950131
+
+#: Length: Meters -> Inches
+M_TO_IN = 39.3701
```
Wouldn't it be better to define these ones as the reciprocal of the below (which are exact values)?

virtuald commented:
I've changed the code to fix some bugs and also process the motion with a 1ms timestep. The motion is a lot better and the robot inertia feels much better than it did previously too!

I updated one of the examples if anyone is interested in playing with this before I finish updating the rest of them: https://github.com/robotpy/examples/tree/tankmodel/physics-4wheel/src

virtuald commented on lib/pyfrc/physics/tankmodel.py:
```diff
@@ -0,0 +1,316 @@
+        
+        robot_mass *= LB_TO_KG
+        
+        self._lmotor = MotorModel(motor_config, l_kv, l_ka, l_vi)
```
Ugh, so the ka/kv have units too. Maybe it's best if we just switch all of the units in this thing over to imperial units. At least it'll be consistent, which is probably more important.

auscompgeek replied:
Well, given that _k<sub>v</sub>_ is voltage over velocity, and _k<sub>a</sub>_ is voltage over acceleration, shouldn't this have been obvious?

Anyway, I vote for SI units. If teams prefer to use imperial units, that should be explicit. There will be teams which will prefer to use SI units, and we shouldn't force them to use imperial units, especially given that the maths here works with SI units.

virtuald replied:
It it were obvious to me, I would have originally considered that as a potential issue. 

I'm torn on this. While SI units are more 'right' -- given that a majority of the teams are in the US, and all of the field units are specified primarily in imperial units, I suspect most FRC high-school users would expect that and be more familiar with that.

Finally, [there's a CD thread for gathering data for this](https://www.chiefdelphi.com/forums/showthread.php?t=161539) and the units there are in US units. I suppose we could convert them.

It would be best to use a units library of some sort so that it didn't matter what users preferred... but in python I have to imagine that would be a performance hit.

virtuald replied:
I think the solution for this (which you alluded to) is to get rid of units entirely, provide motor configurations in both imperial and SI units, and allow the users to use whatever makes sense to them. 

The question then is what units should stall torque be specified in? The SI unit is N-m, but on the CIM datasheet it's oz-in... which doesn't seem like it would be dimensionally correct. I assume the equivalent would be lb-ft, but I'm not able to reproduce the paper's computations for theoretical ka with that number.

virtuald replied:
But from a dimensional point of view, that means that my other units (like robot weight) would need to be in ounces and (wheel diameter) would need to be in inches, right? Whereas if we used lb-ft, then you could stay in feet and pounds.

calcmogul replied:
Depends on how many conversions you want to make during your calculations. You can use any combination of these units.

virtuald replied:
The goal would be to allow users to express things in units that feel natural without lots of conversions.

But, as I said, the numbers don't match the paper whether I use lbf or ozin or whatever, so that needs to be resolved. I posted a question on the paper's CD thread asking which number was used/intended, we'll see if the original author responds.

auscompgeek commented on lib/pyfrc/physics/tankmodel.py:
```diff
@@ -0,0 +1,316 @@
+                
+                k_{v} = \frac{V_{max}}{velocity_{max}}
+                
+                k_{a} = \frac{velocity_{max}}{acceleration_{max}}
```
Rereading the paper, isn't _k<sub>a</sub>_ = _V_<sub>max</sub> / acceleration<sub>max</sub>?

calcmogul replied:
You are correct.

virtuald commented:
Updated and refactored a bunch.

Added the pint units library to do validation and manipulation of various units. This will allow teams to specify the parameters for their motors/robot in SI or fractional units, and the conversions will be done automatically into the correct units.

calcmogul commented on lib/pyfrc/physics/tankmodel.py:
```diff
@@ -0,0 +1,368 @@
+            given mass and robot width/length. If you wish to use a different
+            moment of inertia, set this property after constructing the object
+            
+            Units are ``[length] ** 2 / [mass]``
```
This should be `[mass] * [length] ** 2`.

calcmogul commented on lib/pyfrc/physics/tankmodel.py:
```diff
@@ -0,0 +1,368 @@
+            
+            .. math::
+            
+                velocity_{max} = \frac{\omega_{free} \cdot \pi \cdot d_{wheels} }{r_{gearing}}
```
When introducing variables for the first time in mathematical expressions, one should define them. For example "$d_{wheels}$ is the diameter of the wheels, $\omega_{free}$ is the angular velocity of the wheels with full voltage applied under no load".

virtuald replied:
Thanks for noticing that. Grabbed the definitions from the paper and put them in.

virtuald commented:
One last change before I hit merge. I upgraded the computation from Euler's method to Heun's method, as recommended by Ether's drivetrain paper, and increased the timestep from 0.001 to 0.005. My testing graphs of the output indicate that it's better, even with a timestep of 0.01.

Incidentally, the motion feels better.
